File: client/agent/enroll.py
"""
Called by install.bat: enroll this machine with the server using an enrollment token.
Writes .agent_id and .agent_secret to the agent directory.
"""
import json
import socket
import sys
from pathlib import Path

import requests

def main():
    if len(sys.argv) < 3:
        print("Usage: enroll.py <server_url> <token>")
        sys.exit(1)

    server_url = sys.argv[1].rstrip("/")
    token      = sys.argv[2]

    try:
        import platform, psutil
        ip = psutil.net_if_addrs().get("Ethernet", [{}])[0].get("address", "unknown")
    except Exception:
        ip = "unknown"

    payload = {
        "token":    token,
        "hostname": socket.gethostname(),
        "os":       sys.platform,
        "ip":       ip,
        "username": Path.home().name,
    }

    try:
        resp = requests.post(f"{server_url}/agents/enroll", json=payload, timeout=10)
    except requests.RequestException as e:
        print(f"Connection error: {e}")
        sys.exit(1)

    if not resp.ok:
        print(f"Enrollment failed: {resp.status_code} {resp.text}")
        sys.exit(1)

    data         = resp.json()
    agent_id     = data["agent_id"]
    agent_secret = data["agent_secret"]
    mode         = data.get("mode", "collect_only")

    base = Path(__file__).parent
    (base / ".agent_id").write_text(agent_id,     encoding="utf-8")
    (base / ".agent_secret").write_text(agent_secret, encoding="utf-8")

    config = {
        "server_url": server_url,
        "api_key": "",
        "mode": mode,
        "collection_interval": 10,
        "collect_command_line": True,
        "collect_network": True,
        "anonymize_paths": False,
    }
    (base / "config.json").write_text(json.dumps(config, indent=2), encoding="utf-8")

    print(f"Enrolled as: {agent_id}")
    print(f"Mode: {mode}")

if __name__ == "__main__":
    main()

File: client/agent/src/enroll.c
#include "enroll.h"
#include "sender.h"
#include "config.h"
#include "metadata.h"
#include "json.h"
#include "log.h"
#include <windows.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

/* Simple JSON string value extractor (reused from config.c pattern) */
static int jget(const char *json, const char *key, char *out, int outlen) {
    char search[256];
    snprintf(search, sizeof(search), "\"%s\"", key);
    const char *p = strstr(json, search);
    if (!p) return 0;
    p += strlen(search);
    while (*p == ' ' || *p == '\t') p++;
    if (*p++ != ':') return 0;
    while (*p == ' ' || *p == '\t') p++;
    if (*p++ != '"') return 0;
    int i = 0;
    while (*p && *p != '"' && i < outlen - 1) {
        if (*p == '\\' && *(p + 1)) { p++; }
        out[i++] = *p++;
    }
    out[i] = '\0';
    return 1;
}

int enroll_agent(const char *server_url, const char *token) {
    Metadata meta;
    metadata_collect(&meta);

    JsonBuf jb; jb_init(&jb);
    jb_append(&jb, "{");
    jb_append(&jb, "\"token\":");    jb_str(&jb, token);
    jb_append(&jb, ",\"hostname\":"); jb_str(&jb, meta.hostname);
    jb_append(&jb, ",\"os\":");       jb_str(&jb, meta.os);
    jb_append(&jb, ",\"ip\":");       jb_str(&jb, meta.ip);
    jb_append(&jb, ",\"username\":"); jb_str(&jb, meta.username);
    jb_append(&jb, "}");

    char url[512];
    snprintf(url, sizeof(url), "%s/agents/enroll", server_url);

    int status = 0;
    char *resp = http_post(url, "", jb_get(&jb), &status);
    jb_free(&jb);

    if (!resp) {
        log_error("Enrollment: connection to %s failed", url);
        return 0;
    }
    if (status != 200 && status != 201) {
        log_error("Enrollment: server returned HTTP %d — %s", status, resp);
        free(resp);
        return 0;
    }

    char agent_id[128] = {0}, agent_secret[256] = {0}, mode[32] = "collect_only";
    if (!jget(resp, "agent_id",     agent_id,     sizeof(agent_id))     ||
        !jget(resp, "agent_secret", agent_secret, sizeof(agent_secret))) {
        log_error("Enrollment: missing agent_id/agent_secret in response");
        free(resp);
        return 0;
    }
    jget(resp, "mode", mode, sizeof(mode));
    free(resp);

    config_save_id(agent_id);
    config_save_secret(agent_secret);
    config_write_json(server_url, "", mode);

    log_info("Enrolled as: %s  mode: %s", agent_id, mode);
    return 1;
}
